File: Programmers/BinarySearch/입국심사.py
def binary_search(start, end, target, times):
    count = 0
    if start > end:
        return
    
    mid = int((start+end) // 2)
    for i in range(len(times)):
        count+= mid//times[i]    
    print('mid', mid, 'count', count, 'target', target)
    if count == target:
        for i in times:
            print('moudula', mid%i)
            if mid % i == 0:
                return mid
    count+=1
    if count > target:
        end = mid - 1
    else:
        start = mid + 1        
    return binary_search(start,end,target,times)    

def solution(n, times):
    answer = 0
    data = times.copy()
    
    if n <= len(times):
        answer = max(times)
    else:
        answer = binary_search(0, n*max(data), n, times)
    
    print(answer)
    return answer

File: Programmers/BinarySearch/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

long long solution(int n, vector<int> times) {
    long long answer = 0;
    long long min_time = 1;
    long long max_time = (*max_element(times.begin(), times.end()))*(long long)n;
    long long human;
    
    while(min_time <= max_time){
        human = 0;
        long long avg_time = (min_time+max_time) / 2;
        
        for(auto &t:times){
            human += avg_time/t;
        }       
        if (human >= n) {
            answer = avg_time;
            max_time = avg_time-1;
        }else{
            min_time = avg_time+1;
        }        
    }
    return answer;
}
